lib/puppet/application/device.rb

```diff
@@ -315,10 +315,10 @@ def main
               Puppet.info _("retrieving resource: %{resource} from %{target} at %{scheme}%{url_host}%{port}%{url_path}") % { resource: type, target: device.name, scheme: scheme, url_host: device_url.host, port: port, url_path: device_url.path }
               resources = find_resources(type, name)
               if options[:to_yaml]
-                text = resources.map do |resource|
-                  resource.prune_parameters(:parameters_to_include => @extra_params).to_hierayaml.force_encoding(Encoding.default_external)
-                end.join("\n")
-                text.prepend("#{type.downcase}:\n")
+                data = resources.map do |resource|
+                  resource.prune_parameters(:parameters_to_include => @extra_params).to_hiera_yaml_hash
+                end.inject(:merge)
```

Comment:
nit I think we want `:merge!` here so we don't make multiple copies of the memoized array. Using a simple benchmark, I get:

```
bx ruby inject.rb
                 user     system      total        real
merge!       0.000125   0.000001   0.000126 (  0.000124)
merge        0.006553   0.005585   0.012138 (  0.012142)
```

```ruby
require 'benchmark'

def merge!(array)
  array.inject(:merge!)
end

def merge(array)
  array.inject(:merge)
end

N = 1_000
array = []
0.upto(N) do |i|
  array << { "name#{i}" => { :parameters => { :ensure => :present } } }
end

Benchmark.bm(10) do |x|
  x.report("merge!") { merge!(array) }
  x.report("merge")  { merge(array)  }
end
```

Reply:
Agree, I amended the commit accordingly.  Thanks for this!